### ISSUE TITLE
Optimistic lock not working when used with DateTime objects.

### ISSUE THREAD

**Issue by diego-ninja** [doctrine/orm#8499]
I will try to be as concrete and specific as possible, I think there is a bug when comparing the DateTime objects used in the optimistic lock.

In the UnitOfWork (Line 2497) you are using a !== comparison that only resolves to true when the two instances of the compared object are the same.  This way it is totally impossible for $entityVersion and $lockVersion to be the same at any time since they are different objects. 

```php
    if ($entityVersion !== $lockVersion) {
        throw OptimisticLockException::lockFailedVersionMismatch($entity, $lockVersion, $entityVersion);
    }
```
Debugging to that point I get the following pair of objects, that should pass the validation but they don't:

[Debug variables](https://ibb.co/QXw6kvt)

I have checked the tests that cover the lock functionality and the only thing that is checked is that the appropriate exceptions are thrown, which is always the case since the result of the comparison is always false. There is no test that covers the success of the lock operation.

In older versions of the code the comparison was performed using the != operator with which the functionality works as expected. 

Possible solutions could be to revert to using the != operator or to directly compare the timestamps of DateTime objects using the !== operator.

I can send a PR with the proposed fix.

Thanks in advance.

**Comment by Warxcell** [comment]
Btw there is similar problem with Composite `@Id` that are made of `\DateTimeInterface`

**Comment by diego-ninja** [comment]
Just created a new PR with fix and tests, this time to 2.8.x branch. All checks have passed.
